cf.go

```diff
@@ -0,0 +1,108 @@
+package brokerTags
+
+import (
+	"context"
+	"os"
+
+	"github.com/cloudfoundry-community/go-cfclient/v3/client"
+	"github.com/cloudfoundry-community/go-cfclient/v3/config"
+	"github.com/cloudfoundry-community/go-cfclient/v3/resource"
+)
+
+type CFClientWrapper interface {
+	getOrganizationName(organizationGUID string) (string, error)
+	getServiceInstanceName(instanceGUID string) (string, error)
+	getServiceOfferingName(serviceGUID string) (string, error)
+	getServicePlanName(servicePlanGUID string) (string, error)
+	getSpaceName(spaceGUID string) (string, error)
+}
+
+type OrganizationsInterface interface {
+	Get(ctx context.Context, guid string) (*resource.Organization, error)
+}
+
+type ServiceInstancesInterface interface {
+	Get(ctx context.Context, guid string) (*resource.ServiceInstance, error)
+}
+
+type ServiceOfferingsInterface interface {
+	Get(ctx context.Context, guid string) (*resource.ServiceOffering, error)
+}
+
+type ServicePlansInterface interface {
+	Get(ctx context.Context, guid string) (*resource.ServicePlan, error)
+}
+
+type SpacesInterface interface {
+	Get(ctx context.Context, guid string) (*resource.Space, error)
+}
+
+type cfClientWrapper struct {
+	Organizations    OrganizationsInterface
+	ServiceInstances ServiceInstancesInterface
+	ServiceOfferings ServiceOfferingsInterface
+	ServicePlans     ServicePlansInterface
+	Spaces           SpacesInterface
+}
+
+func NewCFClientWrapper() (*cfClientWrapper, error) {
+	cfg, err := config.NewClientSecret(
+		os.Getenv("CF_API_URL"),
+		os.Getenv("CF_API_CLIENT_ID"),
+		os.Getenv("CF_API_CLIENT_SECRET"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	cf, err := client.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &cfClientWrapper{
+		Organizations:    cf.Organizations,
+		ServiceInstances: cf.ServiceInstances,
+		ServiceOfferings: cf.ServiceOfferings,
+		ServicePlans:     cf.ServicePlans,
+		Spaces:           cf.Spaces,
+	}, nil
+}
+
+func (c *cfClientWrapper) getServiceInstanceName(instanceGUID string) (string, error) {
+	instance, err := c.ServiceInstances.Get(context.Background(), instanceGUID)
```

Comment:
Does it make sense for these to take a `Context`? Do callers typically have a `Context` — if, say, they're handling an HTTP request? If they do, you'll want to accept it so e.g. cancellation works as expected.

Reply:
I don't really understand. You're saying that `getServiceInstanceName` should have a `Context` argument that it passes on?

Reply:
Possibly. When you call it from an actual broker, do you typically have a `Context` available? My guess is yes, because the broker does things in response to HTTP requests, which have an associated `Context`. You'll want the caller to pass their context so if the HTTP request is cancelled, say due to the process being terminated, the caller will call `ctx.Cancel()` and the CF client will be able to respond to that cancellation gracefully.

Reply:
Unfortunately the caller code in `aws-broker` does not make the originating `http.Request`'s Context available to pass in. Possibly it was designed before `context.Context` was introduced in Go ~1.5. We'll keep this as-is for now.